aom/src/main/java/com/nedap/archie/rminfo/SpecificMethodSelector.java

```diff
@@ -0,0 +1,32 @@
+package com.nedap.archie.rminfo;
+
+import java.lang.reflect.Method;
+import java.util.function.BinaryOperator;
+
+/**
+ * Selector which returns the most specific method of two methods.
+ *
+ * This will return the method with the most specific declaring class and return type.
+ *
+ * This will throw an IllegalArgumentException if the methods are incompatible or similar.
+ */
+class SpecificMethodSelector implements BinaryOperator<Method> {
+    public Method apply(Method method1, Method method2) {
+        Class<?> class1 = method1.getDeclaringClass();
+        Class<?> class2 = method2.getDeclaringClass();
+
+        Class<?> returnType1 = method1.getReturnType();
+        Class<?> returnType2 = method2.getReturnType();
+
+        if (class1 == class2 && returnType1 == returnType2) {
+            throw new IllegalArgumentException("Similar methods " + method1.toGenericString() + " and " + method2.toGenericString());
+        } else if (class1.isAssignableFrom(class2) && returnType1.isAssignableFrom(returnType2)) {
+            return method2;
+        } else if (class2.isAssignableFrom(class1) && returnType2.isAssignableFrom(returnType1)) {
+            return method1;
+        } else {
+            // Methods are incompatible
+            throw new IllegalArgumentException("Incompatible methods " + method1.toGenericString() + " and " + method2.toGenericString());
```

Comment:
This will actually trigger on method overloading with different return types (and different method parameters of course). I guess this should not happen for the get/is-methods where this is used for?

The fix is tricky, because you will need to have a method signature as map key, and in openEHR metadata you cannot even have overloading - so I guess not an issue.

Reply:
It currently doesn't get triggered for the AOM and RM, so not an issue at this moment at least.